core/types/rollup_cost.go

```diff
@@ -311,3 +349,125 @@ func l1CostHelper(gasWithOverhead, l1BaseFee, scalar *big.Int) *big.Int {
 	fee.Mul(fee, l1BaseFee).Mul(fee, scalar).Div(fee, oneMillion)
 	return fee
 }
+
+// NewL1CostFuncFjord returns an l1 cost function suitable for the Fjord upgrade
+func NewL1CostFuncFjord(l1BaseFee, l1BlobBaseFee, l1BaseFeeScalar, l1BlobBaseFeeScalar *big.Int) l1CostFunc {
+	return func(costData RollupCostData) (fee, calldataGasUsed *big.Int) {
+		// Fjord L1 cost function:
+		//l1FeeScaled = baseFeeScalar*l1BaseFee*16 + blobFeeScalar*l1BlobBaseFee
+		//estimatedSize = intercept + fastlzCoef*fastlzSize
+		//l1Cost = max(minTransactionSize, estimatedSize) * l1FeeScaled / 1e12
+
+		fastlzSize := new(big.Int).SetUint64(costData.FastLzSize)
+		fastlzSize.Mul(fastlzSize, L1CostFastlzCoef)
+		fastlzSize.Add(fastlzSize, L1CostIntercept)
+
+		if fastlzSize.Cmp(MinTransactionSizeScaled) < 0 {
+			fastlzSize.Set(MinTransactionSizeScaled)
+		}
+
+		calldataGasUsed = new(big.Int).Mul(fastlzSize, new(big.Int).SetUint64(params.TxDataNonZeroGasEIP2028))
+		calldataGasUsed.Div(calldataGasUsed, big.NewInt(1e6))
+
+		calldataCostPerByte := new(big.Int).Mul(l1BaseFee, sixteen)
+		calldataCostPerByte.Mul(calldataCostPerByte, l1BaseFeeScalar)
+		blobCostPerByte := new(big.Int).Mul(l1BlobBaseFee, l1BlobBaseFeeScalar)
+		l1FeeScaled := new(big.Int).Add(calldataCostPerByte, blobCostPerByte)
+
+		l1CostSigned := new(big.Int).Set(fastlzSize)
+		l1CostSigned.Mul(l1CostSigned, l1FeeScaled)
+		l1CostSigned.Div(l1CostSigned, fjordDivisor)
+
+		return l1CostSigned, calldataGasUsed
```

Comment:
nit: The variable naming / reuse is a bit confusing, and the `calldataGasUsed` computation in-between the computation of `l1CostSigned` also threw me off a little while comparing to the specs.